### DISCUSSION TITLE
Marking pull requests as merged

### DISCUSSION THREAD

**Discussion by ArsenArsen** [community/community#6414] [Pull Requests]
Many projects do not merge PRs using GitHubs UI, for various reasons (adding sign-offs, for instance, or signing commits), and instead rely on external automation to track and close PRs after they have been applied. This marks the PRs as closed rather than merged, which has the wrong semantics (closed would imply rejected or withdrawn normally, rather than accepted).

It'd be visually useful to allow automation to mark PRs as merged, leaving closing for rejected PRs, especially of PRs can be marked as merged by a specific person.

An alternative approach could be `Merges: prnum` in commit messages. This would work for most cases I can think of off the top of my head, as merge commits already have commit messages anyway.

**Comment by Wirone** [comment]
Good example of repository with such approach is [`curl/curl`](https://github.com/curl/curl). Recently I had [discussion](https://twitter.com/_Codito_/status/1552393915881472000) about it and [it turned out](https://twitter.com/bagder/status/1552543369829793792) to be the exact case described above. It's misleading for contributors that their PRs are "closed" (rejected) rather than merged. Would be great to be able to set PR state explicitly using UI or Git messages.

**Comment by airtower-luna** [comment]
Peculiar, how did those merges happen? My manual merges (with `git merge`) have always been recognized and closed the relevant PR so far. But I wouldn't expect that to work e.g. with rebasing.

**Reply by Wirone** [reply]
It's described [here](https://github.com/bagder/github-feedback/blob/main/merges.md)

> Since we almost always edit and merge PRs directly in git, without using GitHub, virtually all PRs are just "closed by [commit]" when viewed on GitHub

Those MRs are incorporated into main branch using rebase+squash strategy, but it's done outside of the Github, so these commits can't be recognised as PR's content. Also there is "Closes #PRNUM" in commit message, so Github closes these PRs, but unfortunately it does it in a way that they look like rejected.

**Reply by airtower-luna** [reply]
Yeah, in that case GitHub has no way of knowing. With a proper merge (with fast-forward or a merge commit) the commit IDs are the same and it's possible to detect that all the commits in the PR have been merged.

**Reply by ArsenArsen** [reply]
> Yeah, in that case GitHub has no way of knowing. With a proper merge (with fast-forward or a merge commit) the commit IDs are the same and it's possible to detect that all the commits in the PR have been merged.

Hence the proposal above. These merges aren't really merges, they happen when you apply a patch, and since applying patches has a few advantages over merging (that appear to be recognized by GH, since they also support squash+merge and rebase+merge), I propose that a `Merges:` trailer is supported, or API support, so that some piece of home-grown infra can handle it

**Reply by Wirone** [reply]
@ArsenArsen 

> that appear to be recognized by GH, since they also support squash+merge and rebase+merge

The main difference is that it happens on GH side, so even though Git operations executed on the repo are the same/similar, GH has wider operation context. The same is for Gitlab approvals - when you rebase locally and force push branch, approvals are removed (when project is configured to do so), but when you use `/rebase` command, approvals remain 🙂 It's because Gitlab knows that rebased changes are exactly the same (no amends, no interactive changes). So using "Squash and Merge" in Github's UI also is in safe context, so GH can mark PR as merged.

We just need to be able to mark PR as merged even if rebase+squash was done outside of the GH 🙂

**Reply by OJFord** [reply]
Yes, only if you actually create a merge commit (`git merge --no-ff`). If you amend commits, squash, rebase, or cherry-pick locally then the commit SHAs don't match so (in lieu of something like `Merges #1234` existing like `Closes` in commit messages) GitHub can't know that you effectively merged it, even though squash & rebase 'merges' are supported in the UI.

**Reply by eli-schwartz** [reply]
@OJFord,

Is this some kind of alarming LLM "summary"? As far as I can tell, you're responding to people by telling them what they themselves said several years ago.

**Reply by OJFord** [reply]
> Is this some kind of alarming LLM "summary"?

@eli-schwartz - You're absolutely incorrect!

I was replying to the question since it was # 2 (at least for me) and nothing in the request was clearly explaining why it's important, at least not on this page without reading links elsewhere.

(Vain hope that someone at GH with the resources might see and think _Oh yes, we definitely should handle that_, I guess.)

I'm sure others here understand, it just didn't seem to be stated clearly right here.

**Comment by xmo-odoo** [comment]
Duplicates and older work on the subject:

- #12437 "[Feature Request] Mark PR as Merged from Pull Request API"
- isaacs/github#2 "Explicit "mark as merged" option in pull request UI" (and sibling isaacs/github#548)
- https://forum.bors.tech/t/use-a-squash-merge-in-bors/349/15

**Comment by emaste** [comment]
We would very much like this in FreeBSD as well. Our canonical repo is self-hosted (https://cgit.freebsd.org/src/), but we mirror to GitHub and accept contributions via pull requests to make it easy for contributors who already have GitHub accounts.

We generally add a "Pull Request:" trailer for our own tracking - for example, https://github.com/freebsd/freebsd-src/commit/8117ea0a413150dccc05ea376596b98c4b1855e6 which came from https://github.com/freebsd/freebsd-src/pull/694.

**Comment by Ferdi265** [comment]
I just ran into this as well with a small open source project I run that sees a few contributions once in a blue moon. This behaviour is really confusing and not having any way to tell the git UI that a PR has been manually merged is really unhelpful.

I second adding a "Merges #xyz" tag.

**Comment by mschoettle** [comment]
We are currently migrating some GitLab projects to GitHub (including pull requests). All the migrated PRs (merged and closed) show up as closed which is obviously not great. It would be great to have the ability to manually mark a PR as merged.

**Comment by Diegovsky** [comment]
Yeah I just had to manually solve conflicts in a branch and it ends up being marked as rejected? Github this should be manually overridable!

**Comment by swagtoy** [comment]
I would also sincerely appreciate if `Merges: NNN` was a thing. I occasionally do just like to squash and merge stuff via git and well I dislike having to do `Closes` because it makes it look like the PR was mysteriously closed because I didn't like the person or something. 

Sometimes I do reference the PR number, at least, so it shows there upon close. But it makes it hard to say, "filter" merged pull requests too.
